### PR TITLE
First installment for running w/ replay.

### PR DESCRIPTION
**Pull request by aerorahul**
# Description
This PR:
- does not turn ON replay
- adds functionality to run with replay initial conditions at the end of the assimilation window
- REPLAY can be performed with warm or cold starts.  Cold starts requires an update to the ufs-weather-model past PR  https://github.com/ufs-community/ufs-weather-model/pull/2271
- collects all restart dates that need to be copied to `COM` in a list and then loops over the dates.
- adds functions to create restart and cold start FV3 files so they can be used and not be repeated.

A flag `REPLAY_ICS` should be set to `YES` in `config.base` or `config.fcst` if want to use replay ICS
Running with replay requires perturbation files pretending to be increments as that is the mechanism for generating ensemble ics from deterministic restarts.

# Type of change
- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
Ran forecast jobs on Hera for forecast-only 

# Checklist
- [ ] Any dependent changes have been merged and published
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] I have made corresponding changes to the documentation if necessary

### CODE REVIEW

File: ush/forecast_postdet.sh [hunk 1]
```diff
@@ -84,6 +64,52 @@ FV3_postdet() {
...
       done
+    fi  # if [[ "${RERUN}" != "YES" ]]; then
+
+  fi  # if [[ "${warm_start}" == ".true." ]]; then
+
+  #============================================================================
+  # Determine increment files when doing cold start
+  if [[ "${warm_start}" == ".false." ]]; then
+
+    if [[ "${REPLAY_ICS:-}" == "YES" ]]; then
```
Comment (WalterKolczynski-NOAA) [inline ush/forecast_postdet.sh]: Without an actual default, bash will whine about `==` being used as a unary operator.
```suggestion
    if [[ "${REPLAY_ICS:-NO}" == "YES" ]]; then
```

File: ush/forecast_postdet.sh [hunk 2]
```diff
@@ -121,6 +147,15 @@ EOF
         inc_files=("atminc.nc")
         read_increment=".true."
         res_latlon_dynamics="atminc.nc"
+        if [[ "${REPLAY_ICS:-}" == "YES" ]]; then
```
Comment (WalterKolczynski-NOAA) [inline ush/forecast_postdet.sh]:
```suggestion
        if [[ "${REPLAY_ICS:-NO}" == "YES" ]]; then
```

File: ush/forecast_predet.sh [hunk 3]
```diff
@@ -36,6 +36,40 @@ nhour(){
...
+  tile_files=(gfs_data sfc_data)
+  local nn tt
+  for (( nn = 1; nn <= ntiles; nn++ )); do
+    for tt in "${tile_files[@]}"; do
+      fv3_input_files+=("${tt}.tile${nn}.nc")
+    done
+  done
+  # Create a comma separated string from array using IFS
+  IFS=,
+  echo "${fv3_input_files[*]}"
```
Comment (WalterKolczynski-NOAA) [inline ush/forecast_predet.sh]:
```suggestion
  IFS=, echo "${fv3_input_files[*]}"
```

File: ush/forecast_predet.sh [hunk 4]
```diff
@@ -36,6 +36,40 @@ nhour(){
...
+  tile_files=(fv_core.res fv_srf_wnd.res fv_tracer.res phy_data sfc_data ca_data)
+  local nn tt
+  for (( nn = 1; nn <= ntiles; nn++ )); do
+    for tt in "${tile_files[@]}"; do
+      fv3_restart_files+=("${tt}.tile${nn}.nc")
+    done
+  done
+  # Create a comma separated string from array using IFS
+  IFS=,
+  echo "${fv3_restart_files[*]}"
```
Comment (WalterKolczynski-NOAA) [inline ush/forecast_predet.sh]:
```suggestion
  IFS=, echo "${fv3_restart_files[*]}"
```

File: ush/forecast_predet.sh [hunk 5]
```diff
@@ -54,15 +88,20 @@ common_predet(){
   current_cycle_begin=$(date --utc -d "${current_cycle:0:8} ${current_cycle:8:2} - ${half_window} hours" +%Y%m%d%H)
   current_cycle_end=$(date --utc -d "${current_cycle:0:8} ${current_cycle:8:2} + ${half_window} hours" +%Y%m%d%H)
   next_cycle_begin=$(date --utc -d "${next_cycle:0:8} ${next_cycle:8:2} - ${half_window} hours" +%Y%m%d%H)
-  #Define model start date for current_cycle and next_cycle as the time the forecast will start
+  forecast_end_cycle=$(date --utc -d "${current_cycle:0:8} ${current_cycle:8:2} + ${FHMAX} hours" +%Y%m%d%H)
+
+  # Define model start date for current_cycle and next_cycle as the time the forecast will start
   if [[ "${DOIAU:-}" == "YES" ]]; then
```
Comment (WalterKolczynski-NOAA) [inline ush/forecast_predet.sh]:
```suggestion
  if [[ "${DOIAU:-NO}" == "YES" ]]; then
```

File: ush/forecast_predet.sh [hunk 6]
```diff
@@ -54,15 +88,20 @@ common_predet(){
...
     model_start_date_current_cycle="${current_cycle_begin}"
     model_start_date_next_cycle="${next_cycle_begin}"
   else
-    model_start_date_current_cycle=${current_cycle} 
+    if [[ "${REPLAY_ICS:-}" == "YES" ]]; then
```
Comment (WalterKolczynski-NOAA) [inline ush/forecast_predet.sh]:
```suggestion
    if [[ "${REPLAY_ICS:-NO}" == "YES" ]]; then
```

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

**Comment by aerorahul** [comment]
@NeilBarton-NOAA I should perhaps open this PR against your replay PR.

**Comment by NeilBarton-NOAA** [comment]
@aerorahul that makes more sense in my opinion.

**Comment by aerorahul** [comment]
> @aerorahul that makes more sense in my opinion.



> @aerorahul that makes more sense in my opinion.
Thanks @NeilBarton-NOAA 
I opened one against the branch in your PR
https://github.com/NeilBarton-NOAA/global-workflow/pull/3

**Pull request closed without merging**